### PR TITLE
Fix warmup_bias_lr incorrectly set to 0.0 when auto selects MuSGD

### LINKED ISSUE #23637: Title: optimizer=auto incorrectly sets warmup_bias_lr=0.0 when MuSGD is selected
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

_No response_

### Bug

When optimizer=auto resolves to MuSGD (iterations > 10000), warmup_bias_lr is incorrectly forced to 0.0. This setting is intended for Adam-family optimizers only, but
  currently applies unconditionally to both AdamW and MuSGD.

  Location

  ultralytics/engine/trainer.py, build_optimizer() method:

  # Line 951-952
  name, lr, momentum = ("MuSGD", 0.01, 0.9) if iterations > 10000 else ("AdamW", lr_fit, 0.9)
  self.args.warmup_bias_lr = 0.0  # no higher than 0.01 for Adam  <-- always executes

  The comment says "no higher than 0.01 for Adam", but this line executes regardless of whether MuSGD or AdamW is selected.

  Impact

  - SGD/MuSGD benefits from a non-zero warmup_bias_lr (default 0.1) during warmup to stabilize early training.
  - With warmup_bias_lr=0.0, bias parameters receive zero learning rate at the start of warmup, which can degrade final accuracy.
  - This is not visible in logs because print_args() is called at line 156 (before build_optimizer), so the printed warmup_bias_lr=0.1 does not reflect the actual value used
  during training.
  - Manually specifying optimizer=MuSGD avoids this issue and produces better results with the same lr and momentum.

  To Reproduce

  from ultralytics import YOLO

  model = YOLO("yolo11s-seg.pt")

  # Run 1: auto → MuSGD, warmup_bias_lr silently set to 0.0
  model.train(data="my_data.yaml", epochs=400, optimizer="auto")

  # Run 2: manual MuSGD, warmup_bias_lr stays at 0.1
  model.train(data="my_data.yaml", epochs=400, optimizer="MuSGD")

  # Run 2 produces noticeably higher accuracy than Run 1,
  # despite both using MuSGD with the same lr and momentum.

  Suggested Fix

  Move warmup_bias_lr = 0.0 inside the AdamW branch only:

  if iterations > 10000:
      name, lr, momentum = "MuSGD", 0.01, 0.9
  else:
      name, lr, momentum = "AdamW", lr_fit, 0.9
      self.args.warmup_bias_lr = 0.0  # no higher than 0.01 for Adam

  Environment

  - Ultralytics: 8.4.13
  - Python: 3.11
  - PyTorch: 2.9.1+cu128
  - GPU: NVIDIA A100-PCIE-40GB

  ---

### Environment

(YOLO) root@sx-dms0925:/data/lost+found/sjh/line_seg# yolo check
Ultralytics 8.4.13 🚀 Python-3.11.14 torch-2.9.1+cu128 CUDA:0 (NVIDIA A100-PCIE-40GB, 40444MiB)
Setup complete ✅ (96 CPUs, 754.4 GB RAM, 409.7/438.9 GB disk)

OS                     Linux-5.15.0-156-generic-x86_64-with-glibc2.35
Environment            Linux
Python                 3.11.14
Install                pip
Path                   /root/miniforge3/envs/YOLO/lib/python3.11/site-packages/ultralytics
RAM                    754.35 GB
Disk                   409.7/438.9 GB
CPU                    Intel Xeon Gold 6248R CPU @ 3.00GHz
CPU count              96
GPU                    NVIDIA A100-PCIE-40GB, 40444MiB
GPU count              8
CUDA                   12.8

numpy                  ✅ 2.2.0>=1.23.0
matplotlib             ✅ 3.10.7>=3.3.0
opencv-python          ✅ 4.12.0.88>=4.6.0
pillow                 ✅ 12.0.0>=7.1.2
pyyaml                 ✅ 6.0.3>=5.3.1
requests               ✅ 2.32.5>=2.23.0
scipy                  ✅ 1.16.3>=1.4.1
torch                  ✅ 2.9.1>=1.8.0
torch                  ✅ 2.9.1!=2.4.0,>=1.8.0; sys_platform == "win32"
torchvision            ✅ 0.24.1>=0.9.0
psutil                 ✅ 7.1.3>=5.8.0
polars                 ✅ 1.35.2>=0.20.0
ultralytics-thop       ✅ 2.0.18>=2.0.18

### Minimal Reproducible Example

```
from ultralytics import settings
settings["tensorboard"]=True
from ultralytics import YOLO

from pathlib import Path
import os
import torch
import gc
import sys
import argparse

# from ultralytics import RTDETR


def parse_args():
    parser = argparse.ArgumentParser(description="YOLO Training Script")
    parser.add_argument(
        "--exp_name",
        type=str,
        default=None,
        help="训练实验名称（保存目录名），如 yolo11x_text_v1"
    )
    return parser.parse_args()

if __name__ == '__main__':
    
    args = parse_args()

    # 加载模型
    weights = "./DATA/yolo26s-seg.pt"
    yaml = "ultralytics/cfg/models/26/yolo26-seg.yaml"
    model = YOLO(weights)
    # model = YOLO(yaml).load(weights)  # 从头开始构建新模型
    # model = YOLO("runs/detect/v5n9/weights/best.pt")  # 加载预训练模型（推荐用于训练）
    results = model.train(
        data="DATA/Complete_line/Complete_line_v1.0.yaml",
        epochs=400,
        batch=16,
        imgsz=640,
        workers=14,
        # name=args.exp_name or Path(model.cfg).stem,
        name=args.exp_name or Path(weights).stem, 
        multi_scale=0.25,
        save_period=100,
        # inner_iou=True,
        # inner_iou_version=2,
        # use_elongated_loss=True,
        # project="./runs",
        # device=[6]
        # lr0=0.005,
        optimizer= 'auto' # (str) optimizer: SGD, Adam, Adamax, AdamW, NAdam, RAdam, RMSProp, or auto
    )
```

### Additional

_No response_

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

### PR DESCRIPTION
**Pull request by Br1an67**
Fixes #23637

## Summary

When `optimizer=auto` resolves to MuSGD (iterations > 10000), `warmup_bias_lr` was incorrectly forced to 0.0. This setting is intended for Adam-family optimizers only, but was being applied unconditionally to both AdamW and MuSGD.

SGD/MuSGD benefits from a non-zero `warmup_bias_lr` (default 0.1) during warmup to stabilize early training. With `warmup_bias_lr=0.0`, bias parameters receive zero learning rate at the start of warmup, which can degrade final accuracy.

## Changes

- Modified `build_optimizer()` in `ultralytics/engine/trainer.py` to only set `warmup_bias_lr = 0.0` when AdamW is selected, not when MuSGD is selected
- The conditional logic was restructured from a one-liner to an if-else block to ensure the warmup_bias_lr assignment only happens in the AdamW branch

## Testing

- Verified Python syntax is correct with `ast.parse()`
- The fix preserves the default `warmup_bias_lr = 0.1` when MuSGD is auto-selected
- Manually specifying `optimizer=MuSGD` already avoided this issue and produced better results - now `optimizer=auto` with MuSGD selection behaves the same way

```diff
 ultralytics/engine/trainer.py | 7 +++++--
 1 file changed, 5 insertions(+), 2 deletions(-)
```

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
Fixes an auto-optimizer warmup bug so `warmup_bias_lr` is only forced to `0.0` when `auto` selects `AdamW`, not when it selects `MuSGD` ⚙️

### 📊 Key Changes
- Refactors the `auto` optimizer selection branch in `ultralytics/engine/trainer.py` into explicit `if/else` logic.
- Preserves `MuSGD` auto-selection for larger training runs with `name, lr, momentum = "MuSGD", 0.01, 0.9`.
- Restricts `self.args.warmup_bias_lr = 0.0` to the `AdamW` path only.
- Prevents `MuSGD` runs from inheriting an Adam-specific warmup bias learning rate setting 🛠️

### 🎯 Purpose & Impact
- Fixes unintended optimizer behavior when `auto` chooses `MuSGD`.
- Makes warmup configuration consistent with the selected optimizer and inline code intent.
- Reduces the risk of suboptimal training behavior in long training runs where `MuSGD` is automatically selected 📈
- Improves training reliability with a minimal, targeted Python change.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Br1an67, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! This is an automated message to help with review readiness—an engineer will assist you shortly 😊 Please review the checklist below to help ensure a smooth integration:

-✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
-✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
-✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. If any checks fail, please address the issues.
-✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
-✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
-✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
-✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_  — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing/). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

**Review by UltralyticsAssistant — APPROVED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

The PR looks clean and matches the stated intent. The refactor preserves the existing auto-selection behavior while correctly limiting `warmup_bias_lr = 0.0` to the `AdamW` branch, which fixes the bug without introducing an obvious regression in the shown diff.

**Comment by Br1an67** [comment]
I have read the CLA Document and I sign the CLA

**Comment by glenn-jocher** [comment]
Thanks for digging into the warmup behavior here, @Br1an67. Holding off on this one: the premise is contested by the code owner — @Laughing-q (who authored the auto-optimizer path in #23350/#23234) noted that `warmup_bias_lr = 0.0` on the large-training auto path is intentional legacy behavior, not a bug, and clarified the original semantic misunderstanding. Without a confirmed regression we don't want to change the default learning-rate warmup. If you can share a concrete run where this measurably hurts convergence vs the manual optimizer path, we'd absolutely revisit — please don't hesitate to reopen with data. Thanks! 🙏
